File: menagerie/php-language-signature/generate_assets.php
<?php

declare(strict_types=1);

$root = __DIR__;
$specDir = $root . '/specs';
if (!is_dir($specDir)) {
    mkdir($specDir, 0777, true);
}

$ops = [
    op('source-unit', ['bytes', 'operational_term'], ['String', 'Stmt'], 'Stmt', named([
        ['name' => 'bytes', 'evaluation' => 'unevaluated', 'slot_sort' => 'literal'],
        ['name' => 'operational_term'],
    ]), 'lossless PHP source wrapper; source bytes are recoverable and operational_term is the lifted program'),
    op('seq', ['first', 'second'], ['Stmt', 'Stmt'], 'Stmt', positional(2), 'execute first, then second'),
    op('assign', ['target', 'value'], ['Value', 'Value'], 'Stmt', named([['name' => 'target'], ['name' => 'value']]), 'store value in target'),
    op('if', ['cond', 'then_branch', 'else_branch'], ['Value', 'Stmt', 'Stmt'], 'Stmt', named([['name' => 'cond'], ['name' => 'then_branch'], ['name' => 'else_branch']]), 'conditional branch'),
    op('while', ['cond', 'body'], ['Value', 'Stmt'], 'Stmt', named([['name' => 'cond'], ['name' => 'body']]), 'while loop'),
    op('dowhile', ['body', 'cond'], ['Stmt', 'Value'], 'Stmt', named([['name' => 'body'], ['name' => 'cond']]), 'do-while loop'),
    op('for', ['init', 'cond', 'step', 'body'], ['Stmt', 'Value', 'Stmt', 'Stmt'], 'Stmt', named([['name' => 'init'], ['name' => 'cond'], ['name' => 'step'], ['name' => 'body']]), 'for loop'),
    op('foreach', ['iterable', 'value_name', 'body'], ['Value', 'String', 'Stmt'], 'Stmt', named([['name' => 'iterable'], ['name' => 'value_name', 'slot_sort' => 'identifier'], ['name' => 'body']]), 'foreach loop over value variable'),
    op('return', ['value'], ['Value'], 'Stmt', named([['name' => 'value']]), 'return value'),
    op('break', ['unit'], ['Unit'], 'Stmt', named([['name' => 'unit']]), 'break out of loop'),
    op('continue', ['unit'], ['Unit'], 'Stmt', named([['name' => 'unit']]), 'continue loop'),
    op('throw', ['value'], ['Value'], 'Stmt', named([['name' => 'value']]), 'throw exception or value'),
    op('echo', ['value'], ['Value'], 'Stmt', named([['name' => 'value']]), 'echo value to output'),
    op('print', ['value'], ['Value'], 'Value', named([['name' => 'value']]), 'print value to output'),
    op('exit', ['value'], ['Value'], 'Stmt', named([['name' => 'value']]), 'terminate execution'),
    op('call', ['callee', 'args'], ['String', 'Value'], 'Value', named([['name' => 'callee'], ['name' => 'args', 'shape' => ['kind' => 'set']]]), 'function call'),
    op('methodcall', ['receiver', 'method', 'args'], ['Value', 'String', 'Value'], 'Value', named([['name' => 'receiver'], ['name' => 'method', 'slot_sort' => 'identifier'], ['name' => 'args', 'shape' => ['kind' => 'set']]]), 'instance method call'),
    op('staticcall', ['class', 'method', 'args'], ['String', 'String', 'Value'], 'Value', named([['name' => 'class'], ['name' => 'method', 'slot_sort' => 'identifier'], ['name' => 'args', 'shape' => ['kind' => 'set']]]), 'static method call'),
    op('index', ['base', 'index'], ['Value', 'Value'], 'Value', named([['name' => 'base'], ['name' => 'index']]), 'array offset fetch'),
    op('propfetch', ['receiver', 'property'], ['Value', 'String'], 'Value', named([['name' => 'receiver'], ['name' => 'property', 'slot_sort' => 'identifier']]), 'object property fetch'),
    op('staticprop', ['class', 'property'], ['String', 'String'], 'Value', named([['name' => 'class'], ['name' => 'property', 'slot_sort' => 'identifier']]), 'static property fetch'),
    op('ternary', ['cond', 'then_expr', 'else_expr'], ['Value', 'Value', 'Value'], 'Value', named([['name' => 'cond'], ['name' => 'then_expr', 'evaluation' => 'unevaluated'], ['name' => 'else_expr', 'evaluation' => 'unevaluated']]), 'PHP ternary expression'),
    op('nullcoalesce', ['lhs', 'rhs'], ['Value', 'Value'], 'Value', named([['name' => 'lhs'], ['name' => 'rhs', 'evaluation' => 'unevaluated']]), 'short-circuit null coalescing'),
    op('and', ['lhs', 'rhs'], ['Value', 'Value'], 'Value', named([['name' => 'lhs'], ['name' => 'rhs', 'evaluation' => 'unevaluated']]), 'short-circuit logical conjunction'),
    op('or', ['lhs', 'rhs'], ['Value', 'Value'], 'Value', named([['name' => 'lhs'], ['name' => 'rhs', 'evaluation' => 'unevaluated']]), 'short-circuit logical disjunction'),
    op('not', ['operand'], ['Value'], 'Value', named([['name' => 'operand']]), 'logical negation'),
    op('neg', ['operand'], ['Value'], 'Value', named([['name' => 'operand']]), 'numeric negation'),
    op('pos', ['operand'], ['Value'], 'Value', named([['name' => 'operand']]), 'unary plus'),
    op('bitnot', ['operand'], ['Value'], 'Value', named([['name' => 'operand']]), 'bitwise not'),
];

foreach ([
    'add' => '+', 'sub' => '-', 'mul' => '*', 'div' => '/', 'mod' => '%', 'concat' => '.',
    'eq' => '==', 'ne' => '!=', 'identical' => '===', 'not_identical' => '!==',
    'lt' => '<', 'le' => '<=', 'gt' => '>', 'ge' => '>=',
    'bitand' => '&', 'bitor' => '|', 'bitxor' => '^', 'shl' => '<<', 'shr' => '>>',
] as $name => $symbol) {
    $ops[] = op($name, ['lhs', 'rhs'], ['Value', 'Value'], 'Value', named([['name' => 'lhs'], ['name' => 'rhs']]), "PHP {$symbol} operator");
}

usort($ops, static fn(array $a, array $b): int => $a['file'] <=> $b['file']);
foreach ($ops as $entry) {
    write_json($specDir . '/' . $entry['file'], $entry['spec']);
}

$sortFiles = [];
foreach (['Value', 'Bool', 'Int', 'Real', 'String', 'Unit', 'Stmt'] as $sort) {
    $file = 'sort_' . strtolower($sort) . '.spec.json';
    $sortFiles[] = $file;
    write_json($specDir . '/' . $file, [
        'kind' => 'sort',
        'fn_name' => 'php:sort:' . strtolower($sort),
        'version' => '0.1.0-draft',
        'name' => $sort,
        'locus' => 'menagerie/php-language-signature/README.md#sorts',
    ]);
}

$effects = [
    effect('read', 'Read', ['target'], named([['name' => 'target']]), 'record read from target state cell'),
    effect('write', 'Write', ['target'], named([['name' => 'target']]), 'record write to target state cell'),
    effect('io', 'Io', [], positional(0), 'record IO or network interaction'),
    effect('panic', 'Panic', [], positional(0), 'record throw, die, exit, or fatal trigger_error behavior'),
    effect('unresolved_call', 'Call', ['name'], named([['name' => 'name']]), 'record a call whose callee contract is not available'),
    effect('opaque_loop', 'Loop', ['loopCid'], named([['name' => 'loopCid']]), 'record an opaque loop keyed by its lifted sub-IR CID'),
];
foreach ($effects as $entry) {
    write_json($specDir . '/' . $entry['file'], $entry['spec']);
    write_json($specDir . '/effsig_' . $entry['name'] . '.spec.json', [
        'kind' => 'effect-signature',
        'fn_name' => 'php:effect-signature:' . $entry['name'],
        'version' => '0.1.0-draft',
        'name' => $entry['signature'],
        'locus' => 'menagerie/php-language-signature/README.md#effects',
    ]);
}

write_json($specDir . '/language_signature_php.spec.json', [
    'kind' => 'language_signature',
    'fn_name' => 'php',
    'version' => '0.1.0-draft',
    'sorts' => $sortFiles,
    'operations' => array_map(static fn(array $entry): string => $entry['file'], $ops),
    'equations' => [],
    'effects' => array_map(static fn(array $entry): string => $entry['file'], $effects),
    'effect_signatures' => array_map(static fn(array $entry): string => 'effsig_' . $entry['name'] . '.spec.json', $effects),
    'locus' => 'menagerie/php-language-signature/README.md',
]);

function op(string $name, array $formals, array $sorts, string $returnSort, array $arityShape, string $wp): array
{
    return [
        'file' => 'op_' . $name . '.spec.json',
        'spec' => algorithm('php:' . $name, $formals, $sorts, $returnSort, $name, $arityShape, $wp),
    ];
}

function effect(string $name, string $signature, array $formals, array $arityShape, string $wp): array
{
    $sorts = array_fill(0, count($formals), 'String');
    return [
        'name' => $name,
        'signature' => $signature,
        'file' => 'eff_' . $name . '.spec.json',
        'spec' => algorithm('php:effect:' . str_replace('_', '-', $name), $formals, $sorts, 'Unit', $name . '-effect', $arityShape, $wp, [['kind' => 'effect-signature', 'name' => $signature]]),
    ];
}

function algorithm(string $fnName, array $formals, array $sorts, string $returnSort, string $operator, array $arityShape, string $wp, array $effects = []): array
{
    return [
        'kind' => 'algorithm',
        'fn_name' => $fnName,
        'version' => '0.1.0-draft',
        'formals' => $formals,
        'formal_sorts' => array_map('sort_ctor', $sorts),
        'return_sort' => sort_ctor($returnSort),
        'pre' => ['kind' => 'atomic', 'name' => 'true', 'args' => []],
        'post' => [
            'kind' => 'operation-contract',
            'operator' => $operator,
            'arity' => $sorts,
            'result' => $returnSort,
            'wp' => $wp,
            'arity_shape' => $arityShape,
        ],
        'effects' => ['effects' => $effects],
        'locus' => 'menagerie/php-language-signature/README.md#operations',
    ];
}

function sort_ctor(string $name): array
{
    return ['kind' => 'ctor', 'name' => $name, 'args' => []];
}

function named(array $slots): array
{
    return ['kind' => 'named', 'slots' => $slots];
}

function positional(int $arity): array
{
    return ['kind' => 'positional', 'arity' => $arity];
}

function write_json(string $path, array $value): void
{
    file_put_contents($path, json_encode($value, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n");
}

File: menagerie/ruby-language-signature/generate_assets.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent
SPECS = ROOT / "specs"
VERSION = "0.1.0-draft"
LOCUS = "menagerie/ruby-language-signature/README.md"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--check", action="store_true")
    args = parser.parse_args()

    generated = generate()
    if args.check:
        stale = []
        for rel, value in generated.items():
            path = ROOT / rel
            expected = json_text(value)
            actual = path.read_text(encoding="utf-8") if path.exists() else None
            if actual != expected:
                stale.append(rel)
        if stale:
            raise SystemExit("ruby-language-signature specs are stale: " + ", ".join(stale))
        return

    SPECS.mkdir(parents=True, exist_ok=True)
    for rel, value in generated.items():
        path = ROOT / rel
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json_text(value), encoding="utf-8")


def generate() -> dict[str, dict[str, Any]]:
    files: dict[str, dict[str, Any]] = {}
    sort_names = ["Value", "Bool", "String", "Unit", "Stmt", "ListOfValue"]
    for name in sort_names:
        files[f"specs/sort_{file_key(name)}.spec.json"] = sort_spec(name)

    operations = operation_specs()
    effects = effect_specs()
    effect_sigs = effect_signature_specs()
    files.update({f"specs/{name}": spec for name, spec in operations.items()})
    files.update({f"specs/{name}": spec for name, spec in effects.items()})
    files.update({f"specs/{name}": spec for name, spec in effect_sigs.items()})
    files["specs/language_signature_ruby.spec.json"] = {
        "kind": "language_signature",
        "fn_name": "ruby",
        "version": VERSION,
        "sorts": [f"sort_{file_key(name)}.spec.json" for name in sort_names],
        "operations": list(operations.keys()),
        "equations": [],
        "effects": list(effects.keys()),
        "effect_signatures": list(effect_sigs.keys()),
        "locus": LOCUS,
    }
    return files


def sort_spec(name: str) -> dict[str, Any]:
    return {
        "kind": "sort",
        "fn_name": f"ruby:{name}",
        "formals": [],
        "return_sort": {"kind": "kind", "name": "*"},
        "post": {"kind": "sort-description", "name": name, "description": f"Ruby {name} term sort."},
    }


def operation_specs() -> dict[str, dict[str, Any]]:
    ops: list[tuple[str, list[str], list[str], str, str, dict[str, Any], list[str], str | None]] = [
        (
            "source-unit",
            ["bytes", "operational_term"],
            ["String", "Stmt"],
            "Stmt",
            "lossless Ruby source wrapper; project_effects descends to operational_term",
            named([
                {"name": "bytes", "evaluation": "unevaluated", "slot_sort": "literal"},
                {"name": "operational_term"},
            ]),
            [],
            "The bytes slot carries the original UTF-8 Ruby source as a string.",
        ),
        ("expr", ["value"], ["Value"], "Stmt", "evaluate expression for effects/result", named_names("value"), [], None),
        ("seq", ["first", "second"], ["Stmt", "Stmt"], "Stmt", "wp(first, wp(second, post))", positional(2), [], None),
        ("assign", ["target", "value"], ["Value", "Value"], "Stmt", "store value in target", named_names("target", "value"), ["Writes"], None),
        ("if", ["cond", "then_branch", "else_branch"], ["Bool", "Stmt", "Stmt"], "Stmt", "Ruby if statement", named_names("cond", "then_branch", "else_branch"), [], None),
        ("while", ["cond", "body"], ["Bool", "Stmt"], "Stmt", "opaque Ruby while loop awaiting invariant memento", named_names("cond", "body"), ["OpaqueLoop"], None),
        ("until", ["cond", "body"], ["Bool", "Stmt"], "Stmt", "opaque Ruby until loop awaiting invariant memento", named_names("cond", "body"), ["OpaqueLoop"], None),
        ("for", ["target", "iterable", "body"], ["Value", "Value", "Stmt"], "Stmt", "opaque Ruby for loop over iterable", named_names("target", "iterable", "body"), ["OpaqueLoop"], None),
        ("return", ["value"], ["Value"], "Stmt", "bind function return value", named_names("value"), [], None),
        ("raise", ["exception"], ["Value"], "Stmt", "raise Ruby exception", named_names("exception"), ["Panics"], None),
        (
            "call",
            ["callee", "args"],
            ["String", "ListOfValue"],
            "Value",
            "call callee with positional arguments",
            named([{"name": "callee", "evaluation": "unevaluated", "slot_sort": "literal"}, {"name": "args", "shape": {"kind": "set"}}]),
            ["UnresolvedCall"],
            "The term encoding stores callee followed by positional actual arguments.",
        ),
        (
            "send",
            ["receiver", "method", "args"],
            ["Value", "String", "ListOfValue"],
            "Value",
            "send method to receiver with positional arguments",
            named([
                {"name": "receiver"},
                {"name": "method", "evaluation": "unevaluated", "slot_sort": "literal"},
                {"name": "args", "shape": {"kind": "set"}},
            ]),
            ["UnresolvedCall"],
            "The term encoding stores receiver, method, then positional actual arguments.",
        ),
        ("index", ["base", "index"], ["Value", "Value"], "Value", "Ruby index expression base[index]", named_names("base", "index"), ["Reads"], None),
        ("ivar", ["name"], ["String"], "Value", "instance variable access", literal_name_shape(), ["Reads"], None),
        ("gvar", ["name"], ["String"], "Value", "global variable access", literal_name_shape(), ["Reads"], None),
        ("cvar", ["name"], ["String"], "Value", "class variable access", literal_name_shape(), ["Reads"], None),
        ("const", ["name"], ["String"], "Value", "constant access", literal_name_shape(), ["Reads"], None),
        ("and", ["lhs", "rhs"], ["Bool", "Bool"], "Bool", "short-circuit Ruby conjunction", short_circuit_shape(), [], None),
        ("or", ["lhs", "rhs"], ["Bool", "Bool"], "Bool", "short-circuit Ruby disjunction", short_circuit_shape(), [], None),
        ("not", ["operand"], ["Bool"], "Bool", "boolean negation", named_names("operand"), [], None),
        (
            "ternary",
            ["cond", "then_expr", "else_expr"],
            ["Bool", "Value", "Value"],
            "Value",
            "Ruby cond ? then_expr : else_expr",
            named([
                {"name": "cond"},
                {"name": "then_expr", "evaluation": "unevaluated"},
                {"name": "else_expr", "evaluation": "unevaluated"},
            ]),
            [],
            None,
        ),
    ]

    for name in ["add", "sub", "mul", "div", "mod", "pow", "eq", "ne", "lt", "le", "gt", "ge", "bitand", "bitor", "bitxor", "shl", "shr"]:
        ops.append((name, ["lhs", "rhs"], ["Value", "Value"], "Value", f"Ruby {name} expression", named_names("lhs", "rhs"), [], None))
    for name in ["neg", "pos", "bitnot"]:
        ops.append((name, ["operand"], ["Value"], "Value", f"Ruby unary {name} expression", named_names("operand"), [], None))

    return {
        f"op_{name}.spec.json": op_spec(
            name=name,
            formals=formals,
            formal_sorts=formal_sorts,
            result=result,
            wp=wp,
            shape=shape,
            effect_names=effect_names,
            notes=notes,
        )
        for name, formals, formal_sorts, result, wp, shape, effect_names, notes in ops
    }


def effect_specs() -> dict[str, dict[str, Any]]:
    specs = {
        "reads": ("Reads", ["target"], ["String"], "Value", "read named Ruby state cell"),
        "writes": ("Writes", ["target"], ["String"], "Unit", "write named Ruby state cell"),
        "io": ("IO", ["unit"], ["Unit"], "Unit", "perform Ruby IO"),
        "panics": ("Panics", ["exception"], ["Value"], "Unit", "raise Ruby exception"),
        "unresolved_call": ("UnresolvedCall", ["name"], ["String"], "Value", "call without a resolved contract"),
        "opaque_loop": ("OpaqueLoop", ["loop_cid"], ["String"], "Unit", "loop awaiting invariant memento"),
    }
    return {f"eff_{key}.spec.json": effect_op_spec(key, *value) for key, value in specs.items()}


def effect_signature_specs() -> dict[str, dict[str, Any]]:
    specs = {
        "reads": ("Reads", ["sort_string.spec.json", "sort_value.spec.json"], ["eff_reads.spec.json"]),
        "writes": ("Writes", ["sort_string.spec.json", "sort_unit.spec.json"], ["eff_writes.spec.json"]),
        "io": ("IO", ["sort_unit.spec.json"], ["eff_io.spec.json"]),
        "panics": ("Panics", ["sort_value.spec.json", "sort_unit.spec.json"], ["eff_panics.spec.json"]),
        "unresolved_call": ("UnresolvedCall", ["sort_string.spec.json", "sort_value.spec.json"], ["eff_unresolved_call.spec.json"]),
        "opaque_loop": ("OpaqueLoop", ["sort_string.spec.json", "sort_unit.spec.json"], ["eff_opaque_loop.spec.json"]),
    }
    return {
        f"effsig_{key}.spec.json": {
            "kind": "effect_signature",
            "fn_name": f"ruby:effect-signature:{name}",
            "sorts": sorts,
            "operations": operations,
            "equations": [],
            "effect_signatures": [],
        }
        for key, (name, sorts, operations) in specs.items()
    }


def op_spec(
    *,
    name: str,
    formals: list[str],
    formal_sorts: list[str],
    result: str,
    wp: str,
    shape: dict[str, Any],
    effect_names: list[str],
    notes: str | None,
) -> dict[str, Any]:
    post: dict[str, Any] = {
        "kind": "operation-contract",
        "operator": name,
        "arity": formal_sorts,
        "result": result,
        "wp": wp,
        "arity_shape": shape,
    }
    if notes is not None:
        post["notes"] = notes
    return {
        "kind": "algorithm",
        "fn_name": f"ruby:{name}",
        "formals": formals,
        "formal_sorts": [sort_ctor(sort) for sort in formal_sorts],
        "return_sort": sort_ctor(result),
        "pre": true_formula(),
        "post": post,
        "effects": {"effects": [{"kind": "effect-signature", "name": effect_name} for effect_name in effect_names]},
        "locus": LOCUS,
    }


def effect_op_spec(
    key: str,
    signature_name: str,
    formals: list[str],
    formal_sorts: list[str],
    result: str,
    wp: str,
) -> dict[str, Any]:
    return op_spec(
        name=f"effect:{key}",
        formals=formals,
        formal_sorts=formal_sorts,
        result=result,
        wp=wp,
        shape=named_names(*formals) if formals else positional(0),
        effect_names=[signature_name],
        notes=None,
    )


def sort_ctor(name: str) -> dict[str, Any]:
    return {"kind": "ctor", "name": name, "args": []}


def true_formula() -> dict[str, Any]:
    return {"kind": "atomic", "name": "true", "args": []}


def positional(arity: int) -> dict[str, Any]:
    return {"kind": "positional", "arity": arity}


def named(slots: list[dict[str, Any]]) -> dict[str, Any]:
    return {"kind": "named", "slots": slots}


def named_names(*names: str) -> dict[str, Any]:
    return named([{"name": name} for name in names])


def short_circuit_shape() -> dict[str, Any]:
    return named([{"name": "lhs"}, {"name": "rhs", "evaluation": "unevaluated"}])


def literal_name_shape() -> dict[str, Any]:
    return named([{"name": "name", "evaluation": "unevaluated", "slot_sort": "literal"}])


def file_key(name: str) -> str:
    return name.lower().replace("of", "of_")


def json_text(value: dict[str, Any]) -> str:
    return json.dumps(value, indent=2, ensure_ascii=False) + "\n"


if __name__ == "__main__":
    main()
